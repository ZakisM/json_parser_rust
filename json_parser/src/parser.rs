use bumpalo::{collections::Vec, Bump};

use crate::{
    ast::{JsonProperty, JsonValue},
    error::ExpectedTokenError,
    token::{Lexer, Token, TokenKind},
};

macro_rules! expected_token_err {
    ($actual_token:expr, $row:expr, $column:expr, $expected_token:path) => {
        return Err(ExpectedTokenError::with_offset(
            vec![$expected_token],
            $actual_token.kind,
            ($actual_token.origin).to_owned(),
            $row,
            $column,
        ))
    };
    ($actual_token:expr, $row:expr, $column:expr, $( $variant:ident )|+) => {
        return Err(ExpectedTokenError::with_offset(
            vec![$(TokenKind::$variant),+],
            $actual_token.kind,
            ($actual_token.origin).to_owned(),
            $row,
            $column,
        ))
    };
}

#[derive(Debug)]
pub struct Parser<'a> {
    lexer: Lexer<'a>,
    current_token: Token<'a>,
    peek_token: Token<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut parser = Self {
            lexer: Lexer::new(input),
            current_token: Token::default(),
            peek_token: Token::default(),
        };

        parser.next_token();

        parser
    }

    fn next_token(&mut self) {
        self.current_token = self.peek_token;
        self.peek_token = self.lexer.next_token();
    }

    fn expect_peek(&mut self, expected: TokenKind) -> Result<(), ExpectedTokenError> {
        if self.peek_token.kind != expected {
            expected_token_err!(self.peek_token, self.lexer.row, self.lexer.column, expected)
        }

        self.next_token();

        Ok(())
    }

    fn parse_string(&self, literal: &'a str) -> Result<JsonValue<'a>, ExpectedTokenError> {
        Ok(JsonValue::String(literal))
    }

    fn parse_number(&self, literal: &'a str) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let n = literal.parse::<f64>().map_err(|_| {
            ExpectedTokenError::with_offset(
                vec![TokenKind::Number],
                TokenKind::Illegal,
                literal.to_owned(),
                self.lexer.row,
                self.lexer.column,
            )
        })?;

        Ok(JsonValue::Number(n))
    }

    fn parse_value(&mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let value = match &self.peek_token.kind {
            TokenKind::String => self.parse_string(self.peek_token.origin)?,
            TokenKind::Number => self.parse_number(self.peek_token.origin)?,
            TokenKind::True => JsonValue::Boolean(true),
            TokenKind::False => JsonValue::Boolean(false),
            TokenKind::Null => JsonValue::Null,
            TokenKind::LBrace => self.parse_object(bump)?,
            TokenKind::LBracket => self.parse_array(bump)?,
            _ => {
                expected_token_err!(
                    self.peek_token,
                    self.lexer.row,
                    self.lexer.column,
                    String | Number | True | False | Null | LBrace | LBracket
                )
            }
        };
        self.next_token();

        Ok(value)
    }

    fn parse_property(&mut self, bump: &'a Bump) -> Result<JsonProperty<'a>, ExpectedTokenError> {
        self.expect_peek(TokenKind::String)?;

        let key = self.current_token.origin;

        self.expect_peek(TokenKind::Colon)?;

        let value = self.parse_value(bump)?;

        Ok(JsonProperty::from((key, value)))
    }

    fn parse_array(&mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        self.expect_peek(TokenKind::LBracket)?;

        if self.peek_token.kind == TokenKind::RBracket {
            return Ok(JsonValue::Array(Vec::new_in(bump)));
        }

        let mut items = Vec::with_capacity_in(8, bump);

        loop {
            let value = self.parse_value(bump)?;
            items.push(value);

            match &self.peek_token.kind {
                TokenKind::Comma => self.next_token(),
                TokenKind::RBracket => break,
                _ => {
                    expected_token_err!(
                        self.peek_token,
                        self.lexer.row,
                        self.lexer.column,
                        Comma | RBracket
                    )
                }
            }
        }

        Ok(JsonValue::Array(items))
    }

    fn parse_object(&mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        self.expect_peek(TokenKind::LBrace)?;

        if self.peek_token.kind == TokenKind::RBrace {
            return Ok(JsonValue::Object(Vec::new_in(bump)));
        }

        let mut items = Vec::with_capacity_in(8, bump);

        loop {
            let item = self.parse_property(bump)?;
            items.push(item);

            match &self.peek_token.kind {
                TokenKind::Comma => self.next_token(),
                TokenKind::RBrace => break,
                _ => {
                    expected_token_err!(
                        self.peek_token,
                        self.lexer.row,
                        self.lexer.column,
                        Comma | RBrace
                    )
                }
            }
        }

        Ok(JsonValue::Object(items))
    }

    fn parse_root_object(mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let result = self.parse_object(bump)?;

        self.next_token();

        if !matches!(
            (&self.current_token.kind, &self.peek_token.kind),
            (TokenKind::RBrace, TokenKind::Eof)
        ) {
            expected_token_err!(
                self.peek_token,
                self.lexer.row,
                self.lexer.column,
                TokenKind::Eof
            )
        }

        Ok(result)
    }

    fn parse_root_array(mut self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        let result = self.parse_array(bump)?;

        self.next_token();

        if !matches!(
            (&self.current_token.kind, &self.peek_token.kind),
            (TokenKind::RBracket, TokenKind::Eof)
        ) {
            expected_token_err!(
                self.peek_token,
                self.lexer.row,
                self.lexer.column,
                TokenKind::Eof
            )
        }

        Ok(result)
    }

    pub fn parse(self, bump: &'a Bump) -> Result<JsonValue<'a>, ExpectedTokenError> {
        match self.peek_token.kind {
            TokenKind::LBrace => self.parse_root_object(bump),
            TokenKind::LBracket => self.parse_root_array(bump),
            _ => expected_token_err!(
                self.current_token,
                self.lexer.row,
                self.lexer.column,
                LBrace | LBracket
            ),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bumpalo::vec as bump_vec;

    #[test]
    fn parse_top_level_object() {
        let json = r#"
{
	"string": "Hello, world!",
	"number": -42,
	"nested_object": {
		"nested_string": "This is a nested string",
		"nested_number": [100, 3.21865081787e-6, 300, [400, -500, [600, [700, {"secret": 12345}]]]]
	},
	"boolean": true,
	"nested_deep_empty_array": [[[]], {}],
	"nested_empties": {
	    "empty_object": {},
	    "empty_array": []
	}
}
"#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Ok(JsonValue::Object(bump_vec![in &bump;
                JsonProperty::from(("string", JsonValue::String("Hello, world!"))),
                JsonProperty::from(("number", JsonValue::Number(-42.0))),
                JsonProperty::from((
                    "nested_object",
                    JsonValue::Object(bump_vec![
                        in &bump;
                        JsonProperty::from((
                            "nested_string",
                            JsonValue::String("This is a nested string")
                        )),
                        JsonProperty::from((
                            "nested_number",
                            JsonValue::Array(bump_vec![
                                in &bump;
                                JsonValue::Number(100.0),
                                JsonValue::Number(3.21865081787e-6),
                                JsonValue::Number(300.0),
                                JsonValue::Array(bump_vec![
                                    in &bump;
                                    JsonValue::Number(400.0),
                                    JsonValue::Number(-500.0),
                                    JsonValue::Array(bump_vec![
                                        in &bump;
                                        JsonValue::Number(600.0),
                                        JsonValue::Array(bump_vec![
                                            in &bump;
                                            JsonValue::Number(700.0),
                                            JsonValue::Object(bump_vec![
                                                in &bump;
                                                JsonProperty::from(("secret", JsonValue::Number(12345.0)
                                            ))])
                                        ])
                                    ])
                                ])
                            ])
                        ))
                    ])
                )),
                JsonProperty::from(("boolean", JsonValue::Boolean(true))),
                JsonProperty::from(("nested_deep_empty_array", JsonValue::Array(bump_vec![
                    in &bump;
                    JsonValue::Array(bump_vec![
                        in &bump;
                        JsonValue::Array(bump_vec![
                            in &bump;
                        ])
                    ]),
                    JsonValue::Object(bump_vec![
                        in &bump;
                    ])
                ]))),
                JsonProperty::from(("nested_empties", JsonValue::Object(bump_vec![
                    in &bump;
                    JsonProperty::from(("empty_object", JsonValue::Object(bump_vec![
                        in &bump;
                    ]))),
                    JsonProperty::from(("empty_array", JsonValue::Array(bump_vec![
                        in &bump;
                    ])))
                ]))),
            ]))
        );
    }

    #[test]
    fn parse_top_level_array() {
        let json = r#"
[
    {
        "one": 1,
        "two": 2
    }
]
"#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Ok(JsonValue::Array(bump_vec![
                in &bump;
                JsonValue::Object(bump_vec![
                    in &bump;
                    JsonProperty::from(("one", JsonValue::Number(1.0))),
                    JsonProperty::from(("two", JsonValue::Number(2.0)))
            ])]))
        );
    }

    #[test]
    fn parse_invalid_number() {
        let json = r#"
    {
        "one": 4eee,
        "two": 2
    }
    "#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Err(ExpectedTokenError {
                expected: vec![TokenKind::Number],
                actual: TokenKind::Illegal,
                origin: "4eee".to_owned(),
                row: 3,
                column: 16
            })
        );
    }

    #[test]
    fn parse_invalid_utf8() {
        let json = r#"
    {
        "one": 4èee,
        "two": 2
    }
    "#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Err(ExpectedTokenError {
                expected: vec![TokenKind::Comma, TokenKind::RBrace],
                actual: TokenKind::Illegal,
                origin: "è".to_owned(),
                row: 3,
                column: 17
            })
        );
    }

    #[test]
    fn parse_invalid_octal() {
        let json = r#"
[-65619720000000.61972000000,029]
    "#;

        let bump = Bump::new();
        let parser = Parser::new(json);

        assert_eq!(
            parser.parse(&bump),
            Err(ExpectedTokenError {
                expected: vec![
                    TokenKind::String,
                    TokenKind::Number,
                    TokenKind::True,
                    TokenKind::False,
                    TokenKind::Null,
                    TokenKind::LBrace,
                    TokenKind::LBracket
                ],
                actual: TokenKind::Illegal,
                origin: "0".to_owned(),
                row: 2,
                column: 30
            })
        );
    }
}
