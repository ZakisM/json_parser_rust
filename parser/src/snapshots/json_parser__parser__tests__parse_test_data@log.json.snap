---
source: json_parser/src/parser.rs
expression: parser.parse(&bump)
input_file: test_data/log.json
---
Ok(
    Object(
        [
            JsonProperty {
                key: "jk_host",
                value: String(
                    "jenkinshost100.my-domain-name.net",
                ),
            },
            JsonProperty {
                key: "class_name",
                value: String(
                    "ID\\\\Shared\\\\Logging\\\\Logger",
                ),
            },
            JsonProperty {
                key: "logger_name",
                value: String(
                    "important_logs_go_here!!",
                ),
            },
            JsonProperty {
                key: "cgi_tte_ms",
                value: String(
                    "3173.81",
                ),
            },
            JsonProperty {
                key: "start_timestamp",
                value: Number(
                    1532362756.0,
                ),
            },
            JsonProperty {
                key: "user_agent_device",
                value: String(
                    "Desktop",
                ),
            },
            JsonProperty {
                key: "slush",
                value: String(
                    "facility=important_logs_generated.gelf\\ngelf_version=1.0\\nfile=NA\\nanother_value_includ=205894\\nwe_got_some_one=1\\n@version=1\\nlog_length=1285\\nline=0\\nbools_are_an_awesome_thingy=0\\nuser_agent_version=\\nblubbergnaw_datas=somehostname.and.a.net\\npower2=65360\\nbot_flag=0\\nnumbers=28921\\na_host=somehostname.and.a.net\\nwe_can_urls=https://should.always.be.used.instead.of.http.de/\\nnumargs=42\\nempty_argumt=\\ntype=type01\\nlong_argument_name_here=5\\nanother_bool=0\\nsilly_name_thing=UpperAndLower_Casess\\nyet_anothernother=https://is.much.better.then.http.because.of.stuff/with/path/is/good/too/12345894/",
                ),
            },
            JsonProperty {
                key: "and_an_ip4",
                value: String(
                    "12.345.678.90",
                ),
            },
            JsonProperty {
                key: "@version",
                value: String(
                    "1",
                ),
            },
            JsonProperty {
                key: "error_url_path",
                value: String(
                    "/usr/local/application/with/a/lot/of/sub/directories/in/it.rs",
                ),
            },
            JsonProperty {
                key: "logstash",
                value: String(
                    "who.lines.logstash.probably.on",
                ),
            },
            JsonProperty {
                key: "uuids->",
                value: String(
                    "36ba15e0-d04f-4ae8-a6aa-0d3e65490d76",
                ),
            },
            JsonProperty {
                key: "anotherfilename",
                value: String(
                    "/usr/local/really/long/applicatio/name/that/leads/to/a/script",
                ),
            },
            JsonProperty {
                key: "environment",
                value: String(
                    "Development",
                ),
            },
            JsonProperty {
                key: "floatasstr",
                value: String(
                    "123.56",
                ),
            },
            JsonProperty {
                key: "there_string:",
                value: String(
                    "oh my what?",
                ),
            },
            JsonProperty {
                key: "arry",
                value: Array(
                    [
                        String(
                            "elemen",
                        ),
                        String(
                            "ts ",
                        ),
                        String(
                            "go in",
                        ),
                        String(
                            "her",
                        ),
                        String(
                            "e because I ",
                        ),
                        String(
                            "typ",
                        ),
                        String(
                            "e a",
                        ),
                        String(
                            "lot!!",
                        ),
                    ],
                ),
            },
            JsonProperty {
                key: "message",
                value: String(
                    "This is just a silly json to benchmark to emulate different sturcutres in reality!",
                ),
            },
            JsonProperty {
                key: "argh",
                value: String(
                    "ClbaulgV//Xer8lJ5jr10g==",
                ),
            },
            JsonProperty {
                key: "oh_my_files",
                value: String(
                    "/usr/local/there/are/way/too/many/file/names/in/this/json.txt",
                ),
            },
            JsonProperty {
                key: "user_agent_os",
                value: String(
                    "unknown os",
                ),
            },
            JsonProperty {
                key: "error_host",
                value: String(
                    "even.hosts.have.errors",
                ),
            },
            JsonProperty {
                key: "application",
                value: String(
                    "man",
                ),
            },
            JsonProperty {
                key: "yam_message",
                value: String(
                    "I can just keep writing here for all ethernity and bla bla bla bla it is just more",
                ),
            },
            JsonProperty {
                key: "user_agent_browser",
                value: String(
                    "unknown",
                ),
            },
            JsonProperty {
                key: "error_url",
                value: String(
                    "some.urls.dont.use.https/but/include/a/silly/path/so/we/can/include/them/here.jsonp",
                ),
            },
            JsonProperty {
                key: "short_message",
                value: String(
                    "ARE SHORT BECAUSE!!!",
                ),
            },
            JsonProperty {
                key: "action",
                value: String(
                    "give up.....",
                ),
            },
            JsonProperty {
                key: "cakes!",
                value: String(
                    "please",
                ),
            },
            JsonProperty {
                key: "type",
                value: String(
                    "chocol",
                ),
            },
            JsonProperty {
                key: "log_level",
                value: String(
                    "INFO",
                ),
            },
            JsonProperty {
                key: "too_many_ho",
                value: String(
                    "sts.oh.my.we.can.just.keepit",
                ),
            },
            JsonProperty {
                key: "controller",
                value: String(
                    "no_one_has_controll!",
                ),
            },
            JsonProperty {
                key: "key_keykeykey",
                value: String(
                    "key",
                ),
            },
            JsonProperty {
                key: "a proper_timestamp_ja",
                value: String(
                    "2018-07-23T12:19:16-04:00",
                ),
            },
            JsonProperty {
                key: "and_yet_another",
                value: String(
                    "host.with.lots.of.dots.in.it.gr",
                ),
            },
            JsonProperty {
                key: "@timestamp",
                value: String(
                    "2018-07-23T16:19:16.821Z",
                ),
            },
            JsonProperty {
                key: "level",
                value: Number(
                    6.0,
                ),
            },
        ],
    ),
)
